### PR TITLE
feat: adding certificate delivery subscriptions

### PR DESCRIPTION
**Pull request by Freyskeyd**
# Description

Wire the `broadcast` delivered certificates notification into the new `GraphQL` subscriptions.

An example query: 
Screenshot 2023-10-05 at 15 26 17

A `Websocket` is used to expose and push messages to the frontend.

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

**Review by gruberb — CHANGES_REQUESTED** [review]
👍

**Review by dvdplm — APPROVED** [review]
Left some comments and questions but overall this looks like it's ready to go.

File: crates/topos-tce-api/src/graphql/filter.rs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+pub(crate) enum FilterIs {
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-tce-api/src/graphql/filter.rs:1]: The point is to represent the type of filter used, having just a `bool` wasn't clear.
If you have any suggestion feel free

File: crates/topos-tce-api/src/runtime/mod.rs [hunk 2]
```diff
@@ -185,6 +188,19 @@ impl Runtime {
                     "Dispatching certificate cert_id: {:?} to target subnets: {:?}",
                     &certificate.id, target_subnets
                 );
+
+                // Notify all the transient streams that a new certificate is available
+                // To avoid double allocation for each stream, we clone an Arc of the certificate.
+                // Each stream will convert the UCI certificate into a GraphQL one and send it to the transient stream.
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-tce-api/src/runtime/mod.rs:194]: You mean where is the conversion defined between UCI struct and GraphQL struct? If yes, it's [here](https://github.com/topos-protocol/topos/blob/main/crates/topos-api/src/graphql/certificate.rs#L26)
If you're talking about where we execute the conversion it's [here](https://github.com/topos-protocol/topos/blob/main/crates/topos-tce-api/src/graphql/query.rs#L136)

File: crates/topos-tce-api/src/graphql/query.rs [hunk 3]
```diff
@@ -105,3 +114,78 @@ impl QueryRoot {
...
+                        .unwrap_or(true),
+                )
+            })
+            .map(|c| c.as_ref().into()))
+    }
+}
+
+#[Subscription]
+impl SubscriptionRoot {
+    /// This endpoint is used to received delivered certificates.
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-tce-api/src/graphql/query.rs]: mmh technically you're watching certificates, the `stream` is the transport. I find it less clear to talk about `stream of delivered certificates` as the output of this API call is that we receive `certificate`

**Review by atanmarko — APPROVED** [review]

File: crates/topos-tce-api/src/runtime/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dvdplm) [reply crates/topos-tce-api/src/runtime/mod.rs:193]: 🙏

File: crates/topos-tce-api/src/graphql/filter.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dvdplm) [reply crates/topos-tce-api/src/graphql/filter.rs:1]: How about just `Origin`? Or `OriginFilter`.

**Comment by dvdplm** [comment]
@gruberb Can you re-review please and let us know if you still have concerns?

**Review by gruberb — APPROVED** [review]
👍

**Pull request merged**